### DISCUSSION TITLE
Limit transactions based on sender balance

### DISCUSSION THREAD

**Discussion by BlackTurtle123** [RichardAH/rippled-hooks#43] [Q&A] [answered]
Hello!
Is it possible to write a hook that limits a user to interact with a certain address (with hook), if this user has not enough funds available?

EG; you can only trade with me if you own 100 CUSTOM_TOKEN

**Comment by RichardAH** [comment, marked as answer]
Yes, it is possible.

First obtain the 20 byte account ID of the sender of the originating transaction, and the 20 byte account ID of the issuer of the currency, and construct a buffer containing the 20 byte currency code. Then use `util_keylet` (https://xrpl-hooks.readme.io/reference/util_keylet) with `KEYLET_LINE` to obtain the keylet for the trustline of interest. An example of this is in peggy (https://github.com/XRPL-Labs/xrpld-hooks/blob/e89475e76c9f633e424711bef31b48c7b6d0e51c/hook/examples/peggy/peggy.c#L36-L70)

Next, slot the trustline using the `slot_set` api and the keylet returned from step 1. (https://xrpl-hooks.readme.io/reference/slot_set)
If the trustline between the sender and the issuer for the specified currency code exists then it will be placed into a slot you specified above. If it doesn't exist a negative error code `DOESNT_EXIST` is returned from `slot_set`.

Assuming it exists, you can either dump the whole trustline into a buffer then analyse it inside the hook using `slot` followed by various `sto_` apis, or you can probe it using `slot_subfield`. Hereafter I will explain the latter...

Use `slot_subfield` to (https://xrpl-hooks.readme.io/reference/slot_subfield) to select sfBalance into the same slot. 

Next use `slot` (https://xrpl-hooks.readme.io/reference/slot) to dump the slotted object into a buffer. The slotted object is an STAmount.

Now you have the serialized balance. An easy way to work with this data is `float_sto_set` (https://xrpl-hooks.readme.io/reference/float_sto_set). Use this to turn it into an XFL (https://xrpl-hooks.readme.io/docs/floating-point-numbers-xfl).

Construct a second XFL to compare against using `float_set` (https://xrpl-hooks.readme.io/reference/float_set). The exponent will be 0 and the mantissa 100, if you want to compare against `100.00`.

Finally use `float_compare` (https://xrpl-hooks.readme.io/reference/float_compare) to compare the two XFLs, then use `rollback` (https://xrpl-hooks.readme.io/reference/rollback) if the balance doesn't meet the requirement or `accept` (https://xrpl-hooks.readme.io/reference/accept) if it does.

You should be aware that trustlines have a _high_ and a _low_ account. Which account is which is based on numerical comparison of the 20 byte account IDs (of the sender and issuer). The only difference it will make to you is that the balance you just extracted may be negative. If it is negative you should negate it before performing a comparison using `float_negate` (https://xrpl-hooks.readme.io/reference/float_negate). You can find out if it is negative by doing a `float_compare` against 0. The XFL for 0 is 0.

**Reply by BlackTurtle123** [reply]
So when we have a none standard code like; 0158415500000000C1F76FF6ECB0BAC600000000 (so more then 3 chars and zero's), am I right to assume we convert this into: [1, 88, 65, 85, 26, 116, 138, 210, 193, 247, 111, 246, 236, 176, 204, 205, 0, 0, 0, 0]?
Because I tried it out between following accounts, 
Send from rnVcRbKFYF9aNnKu5XAR97JE1Mx3LodqWA to r3ys7JAc17Hn9YqnLauVJpZxziTBU3W4f, however, I end up with following message: hook: You must have a trustline set

`

      #include "../hookapi.h"
      
      int64_t cbak(int64_t reserved) {
          return 0;
      }
      
      int64_t hook(int64_t reserved ) {

    TRACESTR("Accept.c: Called.");

    uint8_t currency[20] = {1, 88, 65, 85, 26, 116, 138, 210, 193, 247, 111, 246, 236, 176, 204, 205, 0, 0, 0, 0};


    //accept (0,0,0); this would accept every tx automatically

    _g(1,1);   // every hook needs to import guard function and use it at least once

     // get the account the hook is running on
     uint8_t hook_accid[20];
     hook_account(SBUF(hook_accid));

    //get the account  that executed the tx
    uint8_t otxn_accid[20];
    int32_t otxn_accid_len = otxn_field(SBUF(otxn_accid), sfAccount);
    if (otxn_accid_len < 20)
        rollback(SBUF("hook: sfAccount field missing!!!"), 10);
    // unreachable

     // get the source tag if any... negative means it wasn't provided
     int64_t source_tag = otxn_field(0,0, sfSourceTag);
     if (source_tag < 0)
         source_tag = 0xFFFFFFFFU;

     // compare the "From Account" (sfAccount) on the transaction with the account the hook is running on
     int equal = 0; BUFFER_EQUAL(equal, hook_accid, otxn_accid, 20);
     if (equal)
         accept(SBUF("hook: Outgoing transaction"), 20);

     // check if a trustline exists between the sender and the hook for the 7853504543544152000000000000000000000000 currency
     uint8_t keylet[34];
     if (util_keylet(SBUF(keylet), KEYLET_LINE, SBUF(hook_accid), SBUF(otxn_accid), SBUF(currency)) != 34)
         rollback(SBUF("hook: Internal error, could not generate keylet"), 10);

     int64_t user_peggy_trustline_slot = slot_set(SBUF(keylet), 0);
     TRACEVAR(user_peggy_trustline_slot);
     if (user_peggy_trustline_slot < 0)
         rollback(SBUF("hook: You must have a trustline set for token to this account."), 10);

    return 0;
}
`

But there clearly is a trustline set if we look in the explorers

`[
{
account:
"r3ys7JAc17Hn9YqnLauVJpZxziTBU3W4f",
balance:
"8888",
currency:
"7853504543544152000000000000000000000000",
limit:
"10000000000",
limit_peer:
"0",
quality_in:
0,
quality_out:
0
}
]`


Am I missing something very obviously here?

**Reply by RichardAH** [reply]
You should leave it as hex literals like so:
```c++
uint8_t currency[20] = {0x01U, 0x58U, 0x41U, 0x55U, 0x00U, 0x00U, 0x00U, 0x00U, 0xC1U, 0xF7U, 0x6FU, 0xF6U, 0xECU, 0xB0U, 0xBAU, 0xC6U, 0x00U, 0x00U, 0x00U, 0x00U};
```
You are trying to locate a trust line where the issuer is the account the hook is set on and the token holder is the account making the originating transaction. Is this your intention? Or is the issuer a third party? If it is a third party you need to change this code

**Reply by RichardAH** [reply]
Also you have two different currency codes here:
7853504543544152000000000000000000000000
and
0158415500000000C1F76FF6ECB0BAC600000000

**Reply by BlackTurtle123** [reply]
> 7853504543544152000000000000000000000000

You are right I became confused and used the doc example currency code....
In the future the issuer should be a thirdparty yes.
But since hooks are new for me, I am trying to take it step by step.

7853504543544152000000000000000000000000 is what I should have picked.

**Reply by BlackTurtle123** [reply]
I got some follow up questions, since I am pretty new to xrp and hooks.

Do you have docs somewhere about STAmount and what exactly it is?

Also the example in the docs https://xrpl-hooks.readme.io/reference/float_sto_set here doesn't seem to really work, I had to rewrite the rollback with an SBUF and 2 arguments just.

I got following output:
`
    2022-Jun-18 13:25:20.085695481 UTC View:TRC HookTrace[r3ys7JAc17Hn9YqnLauVJpZxziTBU3W4f-rnVcRbKFYF9aNnKu5XAR97JE1Mx3LodqWA]: "Accept.c: Called." 

    2022-Jun-18 13:25:20.085925359 UTC View:TRC HookTrace[r3ys7JAc17Hn9YqnLauVJpZxziTBU3W4f-rnVcRbKFYF9aNnKu5XAR97JE1Mx3LodqWA]: user_peggy_trustline_slot 1

    2022-Jun-18 13:25:20.086064003 UTC View:TRC HookTrace[r3ys7JAc17Hn9YqnLauVJpZxziTBU3W4f-rnVcRbKFYF9aNnKu5XAR97JE1Mx3LodqWA]: sfBalance_slot 2

    2022-Jun-18 13:25:20.086178604 UTC View:TRC HookTrace[r3ys7JAc17Hn9YqnLauVJpZxziTBU3W4f-rnVcRbKFYF9aNnKu5XAR97JE1Mx3LodqWA]: sfBalance_slot 48

    2022-Jun-18 13:25:20.086258480 UTC View:TRC HookTrace[r3ys7JAc17Hn9YqnLauVJpZxziTBU3W4f-rnVcRbKFYF9aNnKu5XAR97JE1Mx3LodqWA]: this should be the balance Float 0*10^(0) <ZERO>

    2022-Jun-18 13:25:20.086331887 UTC View:TRC HookTrace[r3ys7JAc17Hn9YqnLauVJpZxziTBU3W4f-rnVcRbKFYF9aNnKu5XAR97JE1Mx3LodqWA]: this should be min amount Float 8888000000000000*10^(-12)
`

This seems to indicate that balance is 0 or that I made an error in calculating this right?

`

        #include "../hookapi.h"
        
        uint8_t issuer[20] = { // require('ripple-address-codec').decodeAccountID('r3ys7JAc17Hn9YqnLauVJpZxziTBU3W4f')
                0x1U,0x81U,0xb1U,0x98U,0xeeU,0x9fU,0xa2U,0x8bU,0xddU,0xb0U,0x9cU,0xafU,0xa3U,0x86U,0x65U,0x9aU,0xf3U,0x12U,0xf2U,0xadU};
        
        int64_t cbak(int64_t reserved) {
            return 0;
        }
        
        int64_t hook(int64_t reserved ) {
        
            TRACESTR("Accept.c: Called.");
        
            uint8_t currency[20] = {0x78U,0x53U,0x50U,0x45U,0x43U,0x54U,0x41U,0x52U,0x0U,0x0U,0x0U,0x0U,0x0U,0x0U,0x0U,0x0U,0x0U,0x0U,0x0U,0x0U};




    _g(1,1);   // every hook needs to import guard function and use it at least once

     // get the account the hook is running on
     uint8_t hook_accid[20];
     hook_account(SBUF(hook_accid));

    //get the account  that executed the tx
    uint8_t otxn_accid[20];
    int32_t otxn_accid_len = otxn_field(SBUF(otxn_accid), sfAccount);
    if (otxn_accid_len < 20)
        rollback(SBUF("hook: sfAccount field missing!!!"), 10);
    // unreachable

    // get the source tag if any... negative means it wasn't provided
    int64_t source_tag = otxn_field(0,0, sfSourceTag);
    if (source_tag < 0)
        source_tag = 0xFFFFFFFFU;

    // compare the "From Account" (sfAccount) on the transaction with the account the hook is running on
    int equal = 0; BUFFER_EQUAL(equal, hook_accid, otxn_accid, 20);
    if (equal)
        accept(SBUF("hook: Outgoing transaction"), 20);

    // check if a trustline exists between the sender and the hook for the 7853504543544152000000000000000000000000 currency
    uint8_t keylet[34];
    if (util_keylet(SBUF(keylet), KEYLET_LINE, SBUF(issuer), SBUF(otxn_accid), SBUF(currency)) != 34)
        rollback(SBUF("hook: Internal error, could not generate keylet"), 10);

    int64_t user_peggy_trustline_slot = slot_set(SBUF(keylet), 0);
    TRACEVAR(user_peggy_trustline_slot);
    if (user_peggy_trustline_slot < 0)
        rollback(SBUF("hook: You must have a trustline set for token to this account."), 10);

    int64_t sfBalance_slot =
            slot_subfield(user_peggy_trustline_slot, sfBalance, 0);
    TRACEVAR(sfBalance_slot);
    if (sfBalance_slot < 0 )
        rollback(SBUF("sfBalance not found."), 40);


    uint8_t buf[48];
    sfBalance_slot = slot(SBUF(buf), sfBalance_slot);
    TRACEVAR(sfBalance_slot);

    int64_t balance = float_sto_set(sfBalance_slot, 48);
    if (balance < 0)
        rollback(SBUF("Failed to parse serialized float."), 1);
    TRACEVAR(balance);

    int64_t min_amount =
            float_set(0, 8888);

    TRACEVAR(min_amount);


    accept (0,0,0); //this is needed to accept the tx

    return 0;
   }

`

**Reply by BlackTurtle123** [reply]
Another Q; is there some kind of eta for mainnet @RichardAH
